### PR TITLE
Avoid large allocs in mv update code

### PR DESCRIPTION
**Pull request by psarna**
This series addresses #8852 by:
 * migrating to chunked_vector in view update generation code to avoid large allocations
 * reducing the number of futures kept in mutate_MV, tracking how many view updates were already sent

Combined with #8853 I was able to only observe large partition warnings in the logs for the reproducing code, without crashes, large allocation or reactor stall warnings. The reproducing code itself is not part of cql-pytest because I haven't yet figured out how to make it fast and robust.

Tests: unit(release)
Refs  #8852

### CODE REVIEW

File: db/view/view.cc
Comment (avikivity) [inline db/view/view.cc:346]: "The number of view updates can grow large". We won't have a large allocation now with chunked_vector, but what prevents OOM or just extreme memory stress?

File: db/view/view.cc
Comment (avikivity) [inline db/view/view.cc:348]: If the vector can be big, can't this stall?

File: db/view/view.cc
Comment (avikivity) [inline db/view/view.cc:1314]: I think we should proceed even if there is an exception here (but still report an exception), so that we do as much work as we can.

File: db/view/view.cc
Comment (psarna) [reply db/view/view.cc:346]: Nothing at the moment, that's why none of today's patches actually close the issue. Well, maybe this one slightly prevents bad allocs because the allocations are at least in smaller chunks.

File: db/view/view.cc
Comment (psarna) [reply db/view/view.cc:348]: Yeah, it's another candidate for futurizing I guess. Or perhaps this whole machinery should be able to return to results in chunks instead of all at once.

File: db/view/view.cc
Comment (psarna) [reply db/view/view.cc:1210]: I'll try to apply something like that, but in case of this function perhaps a regular semaphore will be enough to keep the number of active continuations below 128.

File: db/view/view.cc
Comment (avikivity) [reply db/view/view.cc:346]: Ok.

File: db/view/view.cc
Comment (avikivity) [reply db/view/view.cc:348]: We need to generate a stream of updates and then process them. I need to revive my idea of supporting

```c++
stream<blah> make_blahs() {
   while (...) {
       co_yield make_a_blah(); 
   }
}
```

and

```c++
future<> eat_blahs(stream<blah> s) {
    while (std::optional<blah> b = co_await s) {
         // do something with *b
    }
}
```

File: db/view/view.cc
Comment (avikivity) [reply db/view/view.cc:1210]: max_concurrent_for_each() is exactly that mix of a semaphore and parallel_for_each.

File: db/view/view.cc
Comment (nyh) [inline db/view/view.cc:1314]: Like Avi said above, I think this is not the right way to limit concurrency. What you do is to split the work into batches of 128 items, and after starting each 128, you wait for all of them. This can lead to partial use of the resources while waiting for stragglers - if for example one out of 100 requests is always twice slower than the rest, half of the time you aren't doing any work. Another problem is that while you wait for the last item of work, the first ones have most likely long finished.
So the right way is to use a semaphore as I suggest in tutorial.md, or Avi said we even added a Seastar primitive to do it for you.

File: db/view/view.cc
Comment (nyh) [reply db/view/view.cc:346]: I'm rusty with this code, so maybe I'm asking for a huge rewrite so feel free to tell me know, but why do we need to prepare all these mutations in a vector (chunked or otherwise) up-front? Why can't we just prepare up to (say) 128 of them and then update each of them, using a semaphore or whatever as proposed in the next patch?

**Review by nyh — CHANGES_REQUESTED** [review]

File: db/view/view.cc
Comment (psarna) [reply db/view/view.cc:346]: It's a small rewrite indeed, my initial idea was to create easily backportable patches that soothe the problem and then think of a way of actually fixing issue #8852. But perhaps it won't be that intrusive, now that I have a fresh morning look over the code. Specifically, this code:
https://github.com/scylladb/scylla/blob/9d49a26e79e675a1296c30dc55cbf023ed6c872f/db/view/view.cc#L901-L923
already has a big `repeat` loop inside it, which would be a natural candidate for returning partial results from the function instead of accumulating all of them in a vector. I think that this series makes sense on its own too, but on top of it I'll try to rework the way view updates are generated so that we don't need to fill the whole vector (chunked or not) in the first place.

**Comment by psarna** [comment]
v2:
 - used `max_concurrent_for_each` as a concurrency-limiting mechanism in mutate_MV

**Comment by psarna** [comment]
Actually, `max_concurrent_for_each` probably still suffers from a problem mentioned before - once any of the updates fail, we'll return the exception and not continue producing updates, while the situation before was that we spawned all updates anyway. I guess I need to rework this loop a bit - when an update fails, the result should be remembered in a helper variable, but then discarded from the main flow in order to let the other updates continue. Only after the whole loop is processed, we should return the exception if it was encountered at least once.

**Comment by avikivity** [comment]
Sure, you can remember the first exception and ignore the rest.

but, what about shutting down? We don't want that to be held up by this loop.

**Comment by psarna** [comment]
I think that the worst that could happen is that the loop would go over all updates, all of which would very quickly fail due to not being able to send anything because the system is shutting down. But I haven't tested it, so I'll take a closer look what happens and maybe add a short-circuit for shutdown.

**Comment by psarna** [comment]
By the way, is there a way of returning an exception from a coroutine with `co_return` without throwing (equivalent of `return make_exception_future<>(std::move(ep))`? My new code uses a coroutine as a glorified do_with in order to keep the container alive, but a very annoying side-effect is that I don't see a way of returning this first encountered exception without rethrowing it :(

**Comment by avikivity** [comment]
> By the way, is there a way of returning an exception from a coroutine with `co_return` without throwing (equivalent of `return make_exception_future<>(std::move(ep))`? My new code uses a coroutine as a glorified do_with in order to keep the container alive, but a very annoying side-effect is that I don't see a way of returning this first encountered exception without rethrowing it :(

No. It's okay to throw, so long as you don't do it very often.

**Comment by psarna** [comment]
> > By the way, is there a way of returning an exception from a coroutine with `co_return` without throwing (equivalent of `return make_exception_future<>(std::move(ep))`? My new code uses a coroutine as a glorified do_with in order to keep the container alive, but a very annoying side-effect is that I don't see a way of returning this first encountered exception without rethrowing it :(
> 
> No. It's okay to throw, so long as you don't do it very often.

My worry is that errors in this layer are usually encountered in packs - e.g. a replica is overloaded and a couple of thousand updates fail. Previous code was able to return an error directly, while the new one would have two things I don't like:
1. instead of `f.failed() + f.get_exception()`, I'd have to use try-catch around a `co_await` twice
2. if any of the updates fail, I'd have to throw instead of returning an exception directly

Due to these two, I'm seriously considering implementing mutate_MV with `do_with` after all.

**Comment by psarna** [comment]
Here's a visualisation of how the new code looks and it sucks: https://github.com/psarna/scylla/commit/f9076ffcc7dd08c28d3766fb22870cc70f7e47d7

**Comment by psarna** [comment]
Turns out I dislike `do_with` even more. v3:
 - reworked mutate_MV to go over all updates and return the first encountered error only after all the updates are tried, as opposed to stopping on the first failure - the original code also spawned all updates in order to perform as much work as possible

File: db/view/view.cc
Comment (nyh) [inline db/view/view.cc:1207]: Food for thought for later (I hope you're keeping track ;-)). Consider that a user deletes 100 partitions in parallel, thinking that 100 is reasonable concurrency. But here each of these deletions causes 128 parallel view updates, so we end up with 12,800 concurrent updates. Maybe we should have a more "global" (per table) semaphore for this view-update concurrency, so its *total* concurrency is limited?

File: db/view/view.cc
Comment (nyh) [inline db/view/view.cc:1323]: I agree that it's sad that after investing so much thinking in the early days of Seastar on having an efficient way to create exceptions *without* throwing, we gave up on this and now we use throw so much.

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
@psarna I want to commit this, but because it doesn't "Fixes" anything, it will be hard to make a decision whether to backport it. Do you think we *should* backport it? If so, maybe you should have at least a Refs to make it easier to find this commit later?

**Comment by psarna** [comment]
GitHub already marks this pull request as referring to #8852 because it's mentioned in the description. But you're right that it would be easier for maintainers if I just explicitly add a "Refs" line in the message, so I'll do that in a second.

**Comment by avikivity** [comment]
> Actually, `max_concurrent_for_each` probably still suffers from a problem mentioned before - once any of the updates fail, we'll return the exception and not continue producing updates, while the situation before was that we spawned all updates anyway. I guess I need to rework this loop a bit - when an update fails, the result should be remembered in a helper variable, but then discarded from the main flow in order to let the other updates continue. Only after the whole loop is processed, we should return the exception if it was encountered at least once.

Does max_concurrent_for_each() really stop processing? I think it's a bug, it should behave like parallel_for_each() and process everything unconditionally.

**Comment by psarna** [comment]
@avikivity you're right, max_concurrent_for_each already does exactly what I mimicked manually in the code. I'll restore the series back to its clearer version, since max_concurrent_for_each already takes care of propagating the exception after all tasks were tried.

**Comment by psarna** [comment]
v4:
 - turned out my assumptions about max_concurrent_for_each were incorrect, which allows me to simplify mutate_MV to the state it was before

File: db/view/view.cc
Comment (psarna) [reply db/view/view.cc:1210]: We do - we have a `_config.view_update_concurrency_semaphore` per table. This limit is just to prevent spawning millions of tasks at once, just to be limited with a semaphore later.

**Comment by avikivity** [comment]
Looks good. I'll let @nyh review again if he wants to.

**Pull request merged**
